common-lib/src/cmr/common/api/web_server.clj

```diff
@@ -55,35 +94,37 @@
   [routes-fn]
   (fn [request]
     (let [^InputStream body-input-stream (:body request)]
-      ;; Init byte array to size MAX_REQUEST_BODY_SIZE + 1
-      (let [buffer-size (inc MAX_REQUEST_BODY_SIZE)
-            input-byte-array (byte-array buffer-size)]
-        (loop [total-bytes-read 0]
-          ;; Read bytes into the byte array up to length MAX_REQUEST_BODY_SIZE + 1. If more is read,
-          ;; we know the request is too large and can return an error
-          (let [bytes-read (.read body-input-stream
-                                  input-byte-array
-                                  total-bytes-read
-                                  (- buffer-size total-bytes-read))
-                at-end? (= -1 bytes-read)
-                ; If bytes-read is -1, still want 0, so when adding it to total bytes doesn't subtract
-                bytes-read (max 0 bytes-read)]
-
-            ;; If the entire request body has been read or if the amount of bytes read is
-            ;; greater than MAX_REQUEST_BODY_SIZE, process based on num bytes read
-            ;; otherwise loop to continue reading the request body
-            (if (or at-end?
-                    (> (+ total-bytes-read bytes-read) MAX_REQUEST_BODY_SIZE))
-              (if (> (+ total-bytes-read bytes-read) MAX_REQUEST_BODY_SIZE)
-                {:status 413
-                 :content-type :text
-                 :body "Request body exceeds maximum size"}
-                ;; Put the request body into a new input stream since the current has been read from
-                (routes-fn (assoc request :body (ByteArrayInputStream.
-                                                 input-byte-array
-                                                 0 ;; offset
-                                                 (+ total-bytes-read bytes-read)))))
-              (recur (+ total-bytes-read bytes-read)))))))))
+      (loop [total-bytes-read 0
+             bytes-read-for-current-buffer 0
+             input-byte-arrays (transient [])
```

Comment:
Using transients because we want calls to add additional buffers to be as fast as possible.